pkg/reconciler/workspace/controller.go

```diff
@@ -306,17 +308,45 @@ func (c *Controller) reconcile(ctx context.Context, workspace *tenancyv1alpha1.W
 		if err != nil {
 			return err
 		}
-		if len(shards) != 0 {
-			targetShard := shards[rand.Intn(len(shards))]
+		// TODO: do we need to handle cross-cluster shard assignment? if so, we need shard name for the shard objects...
+		var filtered []*tenancyv1alpha1.WorkspaceShard
+		for i := range shards {
+			if shards[i].ClusterName == workspace.ClusterName {
```

Comment:
not sure I understand this filtering. It's saying to schedule a Workspace to only those shards that are defined on the same shard the Workspace as object lives on. Why?

Reply:
See the TODO above this - if we want cross-workspace shard assignment OR more than one logical cluster in which shard objects live (maybe it will be just one admin one on the root shard?), then we need to qualify the cluster in the shard assignment. Today, we don't do this - the object only has one string to refer to a shard.

Reply:
Shards should live in the root cluster. If there don't yet, we have to move them there.

Reply:
We can't. It's all written out ... we cannot do that until we have a cross-shard scheduling controller. I need to get this work merged before we can have a virtual workspace that shows one controller all workspaces in need of scheduling.